src/Microsoft.ComponentDetection.Detectors/dotnet/DotNetComponentDetector.cs

```diff
@@ -0,0 +1,186 @@
+        {
+            var namePattern = (projectName ?? "*") + ".dll";
+
+            // look for the compiled output, first as dll then as exe.
+            var candidates = this.directoryUtilityService.EnumerateFiles(projectOutputPath, namePattern, SearchOption.AllDirectories)
+                     .Concat(this.directoryUtilityService.EnumerateFiles(projectOutputPath, namePattern, SearchOption.AllDirectories));
+            foreach (var candidate in candidates)
+            {
+                if (this.IsApplication(candidate))
+                {
+                    return "application";
+                }
+                else
+                {
+                    return "library";
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private bool IsApplication(string assemblyPath)
+    {
+        try
+        {
+            using var peReader = new PEReader(this.fileUtilityService.MakeFileStream(assemblyPath));
+
+            // despite the name `IsExe` this is actually based of the CoffHeader Characteristics
+            return peReader.PEHeaders.IsExe;
+        }
+        catch (Exception)
+        {
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// Recursively get the sdk version from the project directory or parent directories.
+    /// </summary>
+    /// <param name="projectDirectory">Directory to start the search.</param>
+    /// <param name="cancellationToken">Cancellation token to halt the search.</param>
+    /// <returns>Sdk version found, or null if no version can be detected.</returns>
+    private async Task<string?> GetSdkVersionAsync(string projectDirectory, CancellationToken cancellationToken)
+    {
+        // normalize since we need to use as a key
+        projectDirectory = this.pathUtilityService.NormalizePath(projectDirectory);
+        if (this.sdkVersionCache.TryGetValue(projectDirectory, out var sdkVersion))
+        {
+            return sdkVersion;
+        }
+
+        var parentDirectory = this.pathUtilityService.GetParentDirectory(projectDirectory);
+        var globalJsonPath = Path.Combine(projectDirectory, GlobalJsonFileName);
+
+        if (this.fileUtilityService.Exists(globalJsonPath))
+        {
+            var globalJson = await JsonDocument.ParseAsync(this.fileUtilityService.MakeFileStream(globalJsonPath), cancellationToken: cancellationToken);
+            if (globalJson.RootElement.TryGetProperty("sdk", out var sdk))
+            {
+                if (sdk.TryGetProperty("version", out var version))
+                {
+                    sdkVersion = version.GetString();
+                    var globalJsonComponent = new DetectedComponent(new DotNetComponent(sdkVersion));
+                    var recorder = this.ComponentRecorder.CreateSingleFileComponentRecorder(globalJsonPath);
+                    recorder.RegisterUsage(globalJsonComponent, isExplicitReferencedDependency: true);
+                }
+            }
+        }
```

Comment:
I don't think parsing the json file should be the inicial default behavior, we need to always run `dotnet --version` in that location since it can have `latestFeature` rollForward property and SDK values will be higher than declared IIRC. If `dotnet --version` failed, then we can parse it like this.

Reply:
The reason I chose to do the parsing was because I was handling the case where the `dotnet` on the path could not satisfy the `global.json` value.  This is common will be the case for all the `dotnet` repos themselves.

I can add a `dotnet --version` call first, but it could lead to flaky results if the repo doesn't control the environment it runs on completely.

Reply:
I'm OK using `dotnet --version` in this case - we'll just expect that to fail sometimes.

This will mean that in the case of `global.json` we might report something newer than what's listed.  Do you think people might be confused if their global.json says one thing but some repositories are flagged while others (with a newer machine-wide dotnet) are not?

Reply:
CD is meant to flag exactly what was used, so using dotnet --version is a better guess. Likely since in many cases they will bump the SDK in an agent but they won't updated the global.json file.  I think this behavior should be fine but we can validate with users.